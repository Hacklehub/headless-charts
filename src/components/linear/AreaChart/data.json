[
  {
    "year": 2012,
    "macbook": 23.22,
    "services": 12.89,
    "wearables": 10.76,
    "ipad": 30.94,
    "iphone": 78.69
  },
  {
    "year": 2013,
    "macbook": 21.48,
    "services": 16.05,
    "wearables": 10.12,
    "ipad": 31.98,
    "iphone": 91.28
  },
  {
    "year": 2014,
    "macbook": 24.08,
    "services": 18.06,
    "wearables": 8.379,
    "ipad": 30.28,
    "iphone": 101.99
  },
  {
    "year": 2015,
    "macbook": 25.47,
    "services": 19.91,
    "wearables": 10.07,
    "ipad": 23.23,
    "iphone": 155.04
  },
  {
    "year": 2016,
    "macbook": 22.83,
    "services": 24.35,
    "wearables": 11.13,
    "ipad": 20.63,
    "iphone": 136.7
  },
  {
    "year": 2017,
    "macbook": 25.57,
    "services": 32.7,
    "wearables": 12.83,
    "ipad": 18.8,
    "iphone": 139.34
  },

  {
    "year": 2018,
    "macbook": 25.2,
    "services": 39.75,
    "wearables": 17.38,
    "ipad": 18.38,
    "iphone": 164.89
  },
  {
    "year": 2019,
    "macbook": 25.74,
    "services": 46.29,
    "wearables": 24.48,
    "ipad": 21.28,
    "iphone": 142.38
  },
  {
    "year": 2020,
    "macbook": 28.62,
    "services": 53.77,
    "wearables": 30.62,
    "ipad": 23.72,
    "iphone": 137.78
  },

  {
    "year": 2021,
    "macbook": 35.19,
    "services": 68.42,
    "wearables": 38.37,
    "ipad": 31.86,
    "iphone": 191.97
  },
  {
    "year": 2022,
    "macbook": 40.18,
    "services": 78.13,
    "wearables": 41.24,
    "ipad": 29.29,
    "iphone": 205.49
  }
]
