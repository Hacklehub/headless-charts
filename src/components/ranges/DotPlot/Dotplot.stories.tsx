import DotPlot from '.';
import { Meta } from '@storybook/react';

export default {
  title: 'Ranges/Dotplot',
  component: DotPlot,
  tags: ['autodocs'],
} as Meta;

export const Default = {
  args: {
    id: 'dot-plot-default',
    data: [
      //   { set1: 1, set2: 2, set3: 3 },
      //   { set1: 4, set2: 5, set3: 6 },
      //   { set1: 7, set2: 8, set3: 9 },
      { set1: 1 },
      { set2: 2 },
      { set3: 3 },
    ],
    shape: 'circle',
    y: {
      key: 'label',
    },
    x: {
      start: 0,
      end: 1,
      minKey: 'min',
      maxKey: 'max',
      axis: 'bottom',
      axisTicks: 454,
    },
  },
};
