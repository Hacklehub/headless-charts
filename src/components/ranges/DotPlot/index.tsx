import { axisBottom, axisLeft, axisRight, axisTop } from 'd3-axis';
import { defaultChartClassNames, mergeTailwindClasses } from '../../../utils';
import { max, min } from 'd3-array';
import { scaleBand, scaleLinear } from 'd3-scale';
import { select, selectAll } from 'd3-selection';
import {
  symbol,
  symbolCircle,
  symbolCross,
  symbolDiamond,
  symbolSquare,
  symbolStar,
  symbolTriangle,
  symbolWye,
} from 'd3-shape';
/* eslint-disable @typescript-eslint/no-explicit-any */
import { useCallback, useEffect } from 'react';

import { ChartProps } from '../../../types';
import { transition } from 'd3-transition';
import useTooltip from '../../../hooks/useTooltip';
import { zoom } from 'd3-zoom';

interface DotPlotProps extends ChartProps {
  classNameData?: string;
  shape: 'circle';
  y: {
    [key: string]: string;
    axis: 'left' | 'right';
  };
  x: {
    start: number;
    end: number;
    minKey: string;
    maxKey: string;
    axis: 'bottom' | 'top';
    axisTicks: number;
  };
  tooltip?: {
    className?: string;
    keys?: string[];
    html?: (d: any) => string;
  };
}

const DotPlot = ({
  id,
  className,
  data = [],
  classNameData,
  y = { key: 'label', axis: 'left' },
  x,
  margin = {
    top: 40,
    bottom: 40,
    left: 40,
    right: 20,
  },
  padding = {
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
  },
  shape = 'circle',
  tooltip = {},
  zooming,
}: DotPlotProps) => {
  const { onMouseOver, onMouseLeave } = useTooltip({tooltip,
  defaultHtml: (d: any) => `${d[y.key]}: ${d[x.minKey]} to ${d[x.maxKey]}`
  });
  const refreshChart = useCallback(() => {
    const shapeMapping = {
      circle: symbolCircle,
      diamond: symbolDiamond,
      triangle: symbolTriangle,
      square: symbolSquare,
      cross: symbolCross,
      star: symbolStar,
      wye: symbolWye,
    };
    const svg = select(`#${id}`);
    // Clear svg

    svg.selectAll('*').remove();

    const width = +svg.style('width').split('px')[0];
    const height = +svg.style('height').split('px')[0];

    const g = svg.append('g');

    const xFn = scaleLinear()
      /* eslint-disable */
      .domain([
        // @ts-ignore
        Number.isFinite(x.start) ? x.start : min(data.map((d) => d[x.minKey])),
        // @ts-ignore
        Number.isFinite(x.end) ? x.end : max(data.map((d) => d[x.maxKey])),
      ])
      // @ts-ignore
      .range([margin.left, width - padding.right - margin.right]);

    const yFn = scaleBand()
      // @ts-ignore
      .domain(data.map((d) => d[y.key]))
      .range([
        // @ts-ignore
        margin.top + padding.top,
        // @ts-ignore
        height - margin.bottom - padding.bottom,
      ]);

    svg
      .append('clipPath')
      .attr('id', 'clip')
      .append('rect')
      .attr('x', margin.left)
      // @ts-ignore
      .attr('y', margin.top - padding.top - 10)
      .attr('width', width)
      // @ts-ignore
      .attr('height', height + padding.bottom + 8);

    const dataG = g
      .append('g')
      .attr('class', 'data')
      .attr('clip-path', 'url(#clip)');

    const dotRowsG = dataG
      .selectAll('g')
      .data(data)
      .enter()
      .append('g')
      .on(
        'mouseenter',
        // function (event, d) {
        //   tooltip && tooltipDiv.style('opacity', 1);
        //   const [bX, bY] = pointer(event, select('body'));
        //   tooltipDiv.style('left', `${bX + 10}px`).style('top', `${bY + 10}px`);
        //   tooltipDiv.html(
        //     tooltip && tooltip.html
        //       ? tooltip.html(d)
        //       : // @ts-ignore
        //       tooltip.keys
        //       ? // @ts-ignore
        //         tooltip.keys.map((key) => `${key}: ${d[key] || ''}`).join('<br/>')
        //       : // @ts-ignore
        //         `${d[y.key]}: ${d[x.minKey]} to ${d[x.maxKey]}`
        //   );
        // }
        onMouseOver
      )
      .on(
        'mouseleave',
        // function () {
        //   tooltip &&
        //     tooltipDiv
        //       .style('opacity', '0')
        //       .style('left', `0px`)
        //       .style('top', `0px`);
        // }
        onMouseLeave
      );

    transition();

    dotRowsG
      .append('line')
      .attr('clip-path', 'url(#clip)')
      // @ts-ignore
      .attr('x1', (d) => xFn(d[x.minKey]))
      // @ts-ignore
      .attr('x2', (d) => xFn(d[x.minKey]))
      // @ts-ignore
      .attr('y1', (d) => yFn(d[y.key]) + yFn.bandwidth() / 2)
      // @ts-ignore
      .attr('y2', (d) => yFn(d[y.key]) + yFn.bandwidth() / 2)
      .attr('class', `dot-plot-line stroke-current ${classNameData || ``}`)
      .transition()
      .duration(1000)
      // @ts-ignore
      .attr('x2', (d) => xFn(d[x.maxKey]));

    dotRowsG
      .append('path')
      .attr('class', `start-dots fill-current ${classNameData || ''} `)
      .attr('d', () => symbol(shapeMapping[shape], 100)())
      .attr(
        'transform',
        (d) =>
          // @ts-ignore
          `translate(${xFn(d[x.minKey])},${
            // @ts-ignore
            yFn(d[y.key]) + yFn.bandwidth() / 2
          })`
      );

    dotRowsG
      .append('path')
      .attr('class', `end-dots fill-current ${classNameData || ''} `)
      .attr('d', () => symbol(shapeMapping[shape], 100)())
      .attr(
        'transform',
        (d) =>
          // @ts-ignore
          `translate(${xFn(d[x.minKey])},${
            // @ts-ignore
            yFn(d[y.key]) + yFn.bandwidth() / 2
          })`
      )
      .transition()
      .duration(1000)
      .attr(
        'transform',
        (d) =>
          // @ts-ignore
          `translate(${xFn(d[x.maxKey])},${
            // @ts-ignore
            yFn(d[y.key]) + yFn.bandwidth() / 2
          })`
      );

    const redraw = () => {
      selectAll('.dot-plot-line')
        // @ts-ignore
        .attr('x1', (d) => xFn(d[x.minKey]))
        // @ts-ignore
        .attr('x2', (d) => xFn(d[x.maxKey]));

      selectAll('.start-dots').attr(
        'transform',
        (d) =>
          // @ts-ignore
          `translate(${xFn(d[x.minKey])},${
            // @ts-ignore
            yFn(d[y.key]) + yFn.bandwidth() / 2
          })`
      );
      selectAll('.end-dots').attr(
        'transform',
        (d) =>
          // @ts-ignore
          `translate(${xFn(d[x.maxKey])},${
            // @ts-ignore
            yFn(d[y.key]) + yFn.bandwidth() / 2
          })`
      );
    };

    const yAxis = y.axis === 'right' ? axisRight(yFn) : axisLeft(yFn);

    const yAxisG = g
      .append('g')
      .attr('class', 'yAxis axis')
      .attr(
        'transform',
        `translate(${y.axis === 'right' ? margin.left + width : margin.left},0)`
      );

    yAxisG.call(yAxis);

    padding.bottom &&
      yAxisG
        .append('line')
        .attr('x1', 0)
        .attr('x2', 0)
        .attr('y1', margin.top + height - padding.bottom)
        .attr('y2', margin.top + height)
        .attr('stroke', 'currentColor');

    const xAxis =
      x.axis === 'top'
        ? axisTop(xFn).ticks(x.axisTicks || 5)
        : axisBottom(xFn).ticks(x.axisTicks || 5);

    const xAxisG = g.append('g').attr('class', 'axis--x axis ');

    xAxisG
      .attr(
        'transform',
        `translate(0, ${
          x.axis === 'top' ? margin.top : height - margin.bottom
        })`
      )
      .call(xAxis);

    // const tooltipDiv = select('body')
    //   .append('div')
    //   .attr('id', 'tooltip')
    //   .style('position', 'absolute')
    //   .style('opacity', '0')
    //   .attr('class', `tooltip ${(tooltip && tooltip.className) || ''}`);

    if (zooming) {
      const extent = [
        [margin.left, margin.top],
        [width, height],
      ];

      const zoomFunc = zoom()
        .scaleExtent([1, 4])
        // @ts-ignore
        .extent(extent)
        // @ts-ignore
        .translateExtent(extent)
        .on('zoom', zoomed);

      // @ts-ignore
      function zoomed(event) {
        xFn.range(
          // @ts-ignore
          [margin.left, width - padding.right - margin.right].map((d) =>
            event.transform.applyX(d)
          )
        );
        xAxisG.call(xAxis);
        redraw();
      }
      /* eslint-enable */
      svg.call(zoomFunc);
    }
  }, [
    classNameData,
    data,
    id,
    margin,
    padding,
    shape,
    x,
    y,
    zooming,
    onMouseLeave,
    onMouseOver,
  ]);

  useEffect(() => {
    refreshChart();
    return () => {
      selectAll('.tooltip').remove();
    };
  }, [data, refreshChart]);
  return (
    <svg
      id={id}
      className={mergeTailwindClasses(defaultChartClassNames, className)}
    />
  );
};

export default DotPlot;
