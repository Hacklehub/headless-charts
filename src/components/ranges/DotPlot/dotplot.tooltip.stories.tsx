import { Meta, StoryObj } from '@storybook/react';

/* eslint-disable @typescript-eslint/no-explicit-any */
import DotPlot from '.';
import data from './sample.json';

const meta: Meta<typeof DotPlot> = {
  title: 'Ranges/Dotplot/Tooltips',
  component: DotPlot,
  tags: ['autodocs'],
};

export default meta;

type Story = StoryObj<typeof DotPlot>;

export const CustomTooltip: Story = {
  args: {
    data,

    y: {
      key: 'label',
      axis: 'left',
    },
    x: {
      minKey: 'minTemp',
      maxKey: 'maxTemp',
    },
    id: 'dot-plot-custom-tooltip',
    tooltip: {
      keys: ['minTemp', 'maxTemp'],
    },
  },
};

export const CustomTooltipHtml: Story = {
  args: {
    ...CustomTooltip.args,
    id: 'dot-plot-custom-tooltip-html',
    tooltip: {
      html: (d: any) => {
        return `
            <div>
              <div>${d.label}</div>
              <div>Min: ${d.minTemp}</div>
              <div>Max: ${d.maxTemp}</div>
            </div>
          `;
      },
    },
  },
};
