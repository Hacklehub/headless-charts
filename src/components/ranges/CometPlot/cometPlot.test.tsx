import { describe, expect, it } from 'vitest';
import { render, screen } from '@testing-library/react';

import DotPlot from '.';
import data from '../sample.json';

describe('Cometplot', () => {
  it('renders cometPlot component for snapshot test', () => {
    render(
      <DotPlot
        data={data}
        size={40}
        id='cometPlotChart'
        x={{
          fromKey: 'min',
          toKey: 'max',
          className: 'fill-red-800 stoke-white',
          classNameTail: 'fill-white stroke-red-800',
          classNameHead: 'fill-red-800',
        }}
        y={{ key: 'name' }}
        shape='triangle'
      />
    );
    const cometPlot = screen.getByTestId('cometPlot');
    expect(cometPlot).toMatchSnapshot();
  });
  it('renders cometplot component', () => {
    render(
      <DotPlot
        data={data}
        size={40}
        id='cometPlotChart'
        x={{
          fromKey: 'min',
          toKey: 'max',
          className: 'fill-red-800 stoke-white',
          classNameTail: 'fill-white stroke-red-800',
          classNameHead: 'fill-red-800',
        }}
        y={{ key: 'name' }}
        shape='triangle'
      />
    );
    const cometPlot = screen.getByTestId('cometPlot');
    expect(cometPlot).toBeInTheDocument();
  });
});
