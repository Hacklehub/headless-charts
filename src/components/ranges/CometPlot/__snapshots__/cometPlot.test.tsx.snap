// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cometplot > renders cometPlot component for snapshot test 1`] = `
<svg
  class="chart dark:bg-gray-800 dark:text-gray-50 h-64 lg:w-4/12 md:w-6/12 text-gray-900 w-full"
  data-testid="cometPlot"
  id="cometplot"
>
  <g>
    <g
      class="yAxis axis"
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="end"
      transform="translate(40,0)"
    >
      <path
        class="domain"
        d="M-6,40.5H0.5V-39.5H-6"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(0,30)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          Type A
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,10)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          Type B
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,-10)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          Type C
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(0,-30)"
      >
        <line
          stroke="currentColor"
          x2="-6"
        />
        <text
          dy="0.32em"
          fill="currentColor"
          x="-9"
        >
          Type D
        </text>
      </g>
    </g>
    <g
      class="axis--x axis "
      fill="none"
      font-family="sans-serif"
      font-size="10"
      text-anchor="middle"
      transform="translate(0, -40)"
    >
      <path
        class="domain"
        d="M40.5,6V0.5H-19.5V6"
        stroke="currentColor"
      />
      <g
        class="tick"
        opacity="1"
        transform="translate(40.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          0
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(28.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          20
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(16.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          40
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(4.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          60
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(-7.500000000000002,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          80
        </text>
      </g>
      <g
        class="tick"
        opacity="1"
        transform="translate(-19.5,0)"
      >
        <line
          stroke="currentColor"
          y2="6"
        />
        <text
          dy="0.71em"
          fill="currentColor"
          y="9"
        >
          100
        </text>
      </g>
    </g>
    <g
      class="data"
      clip-path="url(#clip)"
    >
      <g>
        <polyline
          class="comet-tail fill-current stroke-0"
          clip-path="url(#clip)"
          points="10 30, 10 25, 10 35, 10 30"
        />
        <path
          class="end-dots fill-current stroke-0"
          d="M5.642,0A5.642,5.642,0,1,1,-5.642,0A5.642,5.642,0,1,1,5.642,0"
          transform="translate(10,30)"
        />
      </g>
      <g>
        <polyline
          class="comet-tail fill-current stroke-0"
          clip-path="url(#clip)"
          points="40 10, 40 5, 40 15, 40 10"
        />
        <path
          class="end-dots fill-current stroke-0"
          d="M5.642,0A5.642,5.642,0,1,1,-5.642,0A5.642,5.642,0,1,1,5.642,0"
          transform="translate(40,10)"
        />
      </g>
      <g>
        <polyline
          class="comet-tail fill-current stroke-0"
          clip-path="url(#clip)"
          points="28 -10, 28 -15, 28 -5, 28 -10"
        />
        <path
          class="end-dots fill-current stroke-0"
          d="M5.642,0A5.642,5.642,0,1,1,-5.642,0A5.642,5.642,0,1,1,5.642,0"
          transform="translate(28,-10)"
        />
      </g>
      <g>
        <polyline
          class="comet-tail fill-current stroke-0"
          clip-path="url(#clip)"
          points="34 -30, 34 -35, 34 -25, 34 -30"
        />
        <path
          class="end-dots fill-current stroke-0"
          d="M5.642,0A5.642,5.642,0,1,1,-5.642,0A5.642,5.642,0,1,1,5.642,0"
          transform="translate(34,-30)"
        />
      </g>
    </g>
  </g>
  <clippath
    id="clip"
  >
    <rect
      height="8"
      width="-40"
      x="40"
      y="30"
    />
  </clippath>
</svg>
`;
